### ISSUE TITLE
"listofvars" Function Failing

### ISSUE THREAD

**Issue by jIyajbe** [maths/moodle-qtype_stack#490]
Issue:
When a student enters an answer in an input text field, the answer validation displays the explicit call to the function stack_validate_listofvars()" instead of simply the list of variables.

Steps to Reproduce:
1. Open an existing STACK question that has "Show the validation" set to "Yes, with variable list" via Preview.
2. Click "Fill in correct responses".

Expected Result:
(version 1) If the answer is a numeric value, show the entered answer with no list (I believe this is the designed behaviour?)

(version 2) If the answer is an algebraic expression, show the entered answer and the list of variables found in it.

Actual Result:
(version 1) Assume the correct answer is 0.145. The system displays:
"Your last answer was interpreted as follows: 0.145
The variables found in your answer were: stack_validate_listofvars(0.145)"

(version 2) Assume the correct answer is (v_f+v_0)/N. The system displays:
"Your last answer was interpreted as follows: vf+v0/N
The variables found in your answer were: stack_validate_listofvars(vf+v0/N)"

NOTES:
1. This issue appeared after upgrading Moodle from 3.6.4 to 3.7, and therefore re-installing STACK (via git). The git pull happened on 28 June 19, approximately 1800 UTC (1100 PST). I am using the existing Maxima image (Linux optimized), from Maxima 5.41. 

2. Healthcheck reports a live connection to the CAS, and does not report any errors.

3. Moodle and Maxima running on macOS 10.13.6 "High Sierra".

4. The text "stack_validate_listofvars() appears to have been rendered by the Tex/MathML engine: {\it stack\_validate\_listofvars}\left(\frac{{v}_{f}+{v}_{0}}{N}\right) See the attached screenshot.

[stack_validate_listofvars() error.pdf](https://github.com/maths/moodle-qtype_stack/files/3340951/stack_validate_listofvars.error.pdf)

**Comment by aharjula** [comment]
Sounds like your update of STACK version included changes in the Maxima side, do test upgrading that image to use the STACK-Maxima scripts from GIT-pull. One should always update the Maxima-image if one updates the STACK-version. The health-check page should tell you if there is a version discrepancy.

Can't be sure though as we do not know what your STACK version was before you did this upgrade. But it is highly likely that you just had a version where those lists were not generated by the Maxima side function `stack_validate_listofvars()` which was added six months ago as a partial fix to issue #432.

**Comment by sangwinc** [comment]
Thanks Matti,

I agree since you said "I am using the existing Maxima image (Linux optimized), from Maxima 5.41.".  Before I investigate further (and I've been using the latest code a lot without mishap like this) please could I ask you to regenerate the Maxima image and confirm if you still have problems?

Chris

**Comment by jIyajbe** [comment]
Hi Matti and Chris,

Well done! I regenerated the image, and the problem disappeared. That's excellent!

On a related note, the automated process for creating an optimised image fails on my machine. Doing it "by hand" worked perfectly, however. Let me know if you would like a formal bug report. (It's worth mentioning that I had two different versions of Maxima installed, in two different paths, so it may be a unique case.)

**Comment by sangwinc** [comment]
Well, STACK can only guess which version you want to use, so if you have two different versions it may not pick up the correct one.  There is a setting to choose Maxima, or you can optimise by hand as you have done.

Any future feedback, suggested improvements (such as this) or contributions very welcome.
Chris

**Comment by timhunt** [comment]
Acutally, STACK does tell you what is going on. The Healthcheck page (linked from Admin -> Plugins -> Question types -> STACK includes a check that you have the right version of the Maxima code being used. You should always check that when upgrading STACK.

**Comment by jIyajbe** [comment]
Oh yes, the healthcheck definitely explained the version conflicts. The issue is (was) that when I had two different Maximas (5.38.0 and 5.41) installed. To set things up, I used “Linux” (i.e., not an optimised image),  and I put the explicit path to 5.41 in the STACK settings, and set the expected version (the drop-down list) to 5.41, and that worked perfectly. From that, I clicked “Create Maxima image” (I forget the exact text on that button), but that failed; the healthcheck page reported a Maxima version mismatch. Something like “Version 5.41 expected, but found 5.38.0”. Deleting 5.38.0 from my system didn’t change that result.

But, when I created the optimised image by hand, it worked perfectly.
